### PR TITLE
Ensure the metadata object gets written correctly

### PR DESCRIPTION
**Pull request by chadhietala**
Now uses Filter's cache to compute the evicted files so that dep-graph.json does not go stale.

### CODE REVIEW

**Comment by stefanpenner** [comment]
lets not merge this, rather try to fix it upstream: https://github.com/caitp/cauliflower-filter/issues/12

File: index.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ var path       = require('path');
 var fs         = require('fs');
 var stringify  = require('json-stable-stringify');
 var mergeTrees = require('broccoli-merge-trees');
-var funnel = require('broccoli-funnel');
```
Comment (stefanpenner) [inline index.js]: do we need this?

File: index.js [hunk 2]
```diff
@@ -7,7 +7,7 @@ var path       = require('path');
...
+var funnel     = require('broccoli-funnel');
```
Comment (stefanpenner) [inline index.js]: do we need this ?

File: index.js [hunk 3]
```diff
@@ -63,13 +63,26 @@ Babel.prototype.rebuild = function() {
...
     if (self.exportModuleMetadata) {
-      fs.writeFileSync(self.outputPath + path.sep + 'dep-graph.json', stringify(self.moduleMetadata, {
-        space: 2
-      }));
+      self.generateDepGraph();
     }
   });
 };
 
+Babel.prototype.generateDepGraph = function() {
```
Comment (stefanpenner) [inline index.js]: prefix this with `_` since its private.

File: index.js [hunk 4]
```diff
@@ -63,13 +63,26 @@ Babel.prototype.rebuild = function() {
...
+  var imports = Object.keys(this.moduleMetadata);
+  var evictedImports = diff(imports, residentImports);
+
+  if (evictedImports.length > 0) {
+    evictedImports.forEach(function(importName) {
+      delete this.moduleMetadata[importName];
+    }, this);
+  }
+
+  fs.writeFileSync(this.outputPath + path.sep + 'dep-graph.json',
```
Comment (stefanpenner) [inline index.js]: should we skip this if nothing changed?

File: index.js (hunk 2 above)
Comment (chadhietala) [inline index.js]: See https://github.com/babel/broccoli-babel-transpiler/blob/master/index.js#L51. I just aligned it.

Comment (stefanpenner) [inline index.js]: ah, carry on

File: index.js (hunk 4 above)
Comment (chadhietala) [inline index.js]: Yup.

Comment (chadhietala) [inline index.js]: Ah actually you need to always write it as it's a sideeffect file I believe.

**Comment by chadhietala** [comment]
Ah actually I believe you need to always write it as it's a side effect file. It's computed as a result of a build, if you don't always write it will be missing from the idempotent rebuild trees.

**Comment by stefanpenner** [comment]
Okidokies

**Pull request merged**